libraries/c_sdk/aws/defender/test/unit/aws_iot_tests_defender_unit.c

```diff
@@ -0,0 +1,295 @@
+    _mockedMqttConnection = true;
+    _startInfo.mqttConnection = _mqttConnection;
+
+    AwsIotDefenderError_t error = AwsIotDefender_Start( &_startInfo );
+
+    TEST_ASSERT_EQUAL( AWS_IOT_DEFENDER_SUCCESS, error );
+
+    /* Set "all metrics" for TCP connections metrics group. */
+    error = AwsIotDefender_SetMetrics( AWS_IOT_DEFENDER_METRICS_TCP_CONNECTIONS,
+                                       AWS_IOT_DEFENDER_METRICS_ALL );
+
+    TEST_ASSERT_EQUAL( AWS_IOT_DEFENDER_SUCCESS, error );
+
+    TEST_ASSERT_EQUAL( AWS_IOT_DEFENDER_METRICS_ALL, _AwsIotDefenderMetrics.metricsFlag[ AWS_IOT_DEFENDER_METRICS_TCP_CONNECTIONS ] );
+}
+
+TEST( Defender_Unit, Start_with_invalid_mqtt_connection )
+{
+    /* Set uninitialized MQTT connection */
+    _startInfo.mqttConnection = IOT_MQTT_CONNECTION_INITIALIZER;
+
+    AwsIotDefenderError_t error = AwsIotDefender_Start( &_startInfo );
+
+    TEST_ASSERT_EQUAL( AWS_IOT_DEFENDER_INVALID_INPUT, error );
+}
+
+TEST( Defender_Unit, Start_should_return_success )
+{
+    /* Set up a mocked MQTT connection. */
+    TEST_ASSERT_EQUAL_INT( true, IotTest_MqttMockInit( &_mqttConnection ) );
+    _mockedMqttConnection = true;
+    _startInfo.mqttConnection = _mqttConnection;
+
+    AwsIotDefenderError_t error = AwsIotDefender_Start( &_startInfo );
+
+    TEST_ASSERT_EQUAL( AWS_IOT_DEFENDER_SUCCESS, error );
+}
+
+TEST( Defender_Unit, Start_should_return_err_if_already_started )
+{
+    /* Set up a mocked MQTT connection. */
+    TEST_ASSERT_EQUAL_INT( true, IotTest_MqttMockInit( &_mqttConnection ) );
+    _mockedMqttConnection = true;
+    _startInfo.mqttConnection = _mqttConnection;
+
+    AwsIotDefenderError_t error = AwsIotDefender_Start( &_startInfo );
+
+    TEST_ASSERT_EQUAL( AWS_IOT_DEFENDER_SUCCESS, error );
+
+    /* Start defender for a second time. */
+    error = AwsIotDefender_Start( &_startInfo );
+
+    TEST_ASSERT_EQUAL( AWS_IOT_DEFENDER_ALREADY_STARTED, error );
+}
+
+TEST( Defender_Unit, SetPeriod_too_short )
+{
+    TEST_ASSERT_EQUAL( AWS_IOT_DEFENDER_PERIOD_TOO_SHORT, AwsIotDefender_SetPeriod( AWS_IOT_DEFENDER_DEFAULT_PERIOD_SECONDS - 1 ) );
+}
+
+TEST( Defender_Unit, SetPeriod_with_proper_value )
+{
+    TEST_ASSERT_EQUAL( AWS_IOT_DEFENDER_SUCCESS, AwsIotDefender_SetPeriod( AWS_IOT_DEFENDER_DEFAULT_PERIOD_SECONDS + 1 ) );
+
+    TEST_ASSERT_EQUAL( AWS_IOT_DEFENDER_DEFAULT_PERIOD_SECONDS + 1, AwsIotDefender_GetPeriod() );
+}
+
+TEST( Defender_Unit, SetPeriod_after_started )
+{
+    /* Set up a mocked MQTT connection. */
+    TEST_ASSERT_EQUAL_INT( true, IotTest_MqttMockInit( &_mqttConnection ) );
+    _mockedMqttConnection = true;
+    _startInfo.mqttConnection = _mqttConnection;
+
+    TEST_ASSERT_EQUAL( AWS_IOT_DEFENDER_SUCCESS,
+                       AwsIotDefender_Start( &_startInfo ) );
+
+    TEST_ASSERT_EQUAL( AWS_IOT_DEFENDER_SUCCESS, AwsIotDefender_SetPeriod( 600 ) );
+
+    TEST_ASSERT_EQUAL( 600, AwsIotDefender_GetPeriod() );
```

Comment:
Might as well create a macro for 600 too. Or use `2 * AWS_IOT_DEFENDER_DEFAULT_PERIOD_SECONDS`

Reply:
Good catch